src/repositories/updater.rs

```
@@ -0,0 +1,366 @@
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(updater) = GitLab::new("gitlab.freedesktop.org", &None) {
+            updaters.push(Box::new(updater));
+        }
+        Self { updaters, pool }
+    }
+
+    pub(crate) fn load_repository(&self, metadata: &MetadataPackage) -> Result<Option<i32>> {
+        let url = match &metadata.repository {
+            Some(url) => url,
+            None => {
+                debug!("did not collect stats as no repository URL was present");
+                return Ok(None);
+            }
+        };
+        let mut conn = self.pool.get()?;
+        self.load_repository_inner(&mut conn, url)
+    }
+
+    fn load_repository_inner(&self, conn: &mut Client, url: &str) -> Result<Option<i32>> {
+        let name = match repository_name(url) {
+            Some(name) => name,
+            None => return Ok(None),
+        };
+
+        // Avoid querying the APIs for repositories we already loaded.
+        if let Some(row) = conn.query_opt(
+            "SELECT id FROM repositories WHERE name = $1 AND host = $2 LIMIT 1;",
+            &[&format!("{}/{}", name.owner, name.repo), &name.host],
+        )? {
+            return Ok(Some(row.get("id")));
+        }
+        for updater in self.updaters.iter().filter(|u| u.host() == name.host) {
+            let res = match updater.fetch_repository(&name) {
+                Ok(Some(repo)) => self.store_repository(conn, updater.host(), repo),
+                Err(err) => {
+                    warn!("failed to collect `{}` stats: {}", updater.host(), err);
+                    return Ok(None);
+                }
+                _ => continue,
+            };
+            return match res {
+                Ok(repo_id) => Ok(Some(repo_id)),
+                Err(err) => {
+                    warn!("failed to collect `{}` stats: {}", updater.host(), err);
+                    Ok(None)
```

Why return Ok(None) here when you return the error above?

---

Because I already displayed a warning. Can be changed, as you prefer.

---

Same as above, I'd prefer to return an error.